### ISSUE TITLE
Question: Qwen3-30B-A3B throughput difference between PithTrain EP4/PP2 and TorchTitan FSDP8 on 8×B200

### ISSUE THREAD

**Issue by DAN-DNA-AND** [mlc-ai/pith-train#79]
Hi PithTrain team,

Thank you for releasing this work. We reproduced the official Qwen3-30B-A3B seq8192 EP4/PP2 recipe on one 8×B200 node and observed approximately 17.5k tokens/s/GPU.

For comparison, we tested TorchTitan using the same checkpoint, DCLM token stream, tokenizer, causal-LM objective, and router-aux coefficient:

- BF16, seq8192
- FSDP8, EP1, PP1
- MBS8, GBS64
- Selective activation checkpointing
- Eight loss chunks
- Compiled Transformer blocks and loss

Its stable steps reached 19.41k, 19.53k, and 19.52k tokens/s/GPU, with 141.79 GiB peak memory per GPU.
We understand that the parallelism and batch geometry differ, and this was only a five-step comparison. Is this performance difference expected, or are there additional PithTrain settings we should match for a fairer comparison?

I assume EP may become more advantageous at larger scales, but I would greatly appreciate your perspective.

Thank you for your time!

**Comment by haok1402** [comment]
@DAN-DNA-AND Thanks for bringing this to our attention! Would you mind sharing the scripts you used on both frameworks so we can try reproducing on our end? Also, while using the same checkpoint, what's the reported load-balancing loss here?

**Comment by DAN-DNA-AND** [comment]
@haok1402 # 
Thanks for the reply. I would be happy to share the reproduction details.

I expect to make the TorchTitan implementation that available within 1 day and will
post the link here as soon as it is ready.

For the PithTrain side, we used the upstream repository without tracked source
changes, pinned to this exact commit:

```text
23db1829a2e55596a5347a94e53a770721e79539
```


The following is the configuration used for our seq8192 PithTrain run. The
paths below should point to PithTrain's standard Qwen3 model configuration,
prepared DCLM dataset, and Qwen3-30B-A3B checkpoint conversion output.

```python
from pathlib import Path

from pithtrain.modules.logging import LoggingWandbCfg
from pithtrain.tasks.pretrain_language_model import (
    PretrainLanguageModelCfg,
    launch,
)


cfg = PretrainLanguageModelCfg()

# One 8xB200 node: PP2 x EP4, matching the paper recipe.
cfg.distributed.context_parallel_size = 1
cfg.distributed.pipeline_parallel_size = 2
cfg.distributed.expert_parallel_size = 4

training = cfg.training
training.model = Path(
    "examples/pretrain_language_model/qwen3-30b-a3b/config.json"
).resolve()
training.dataset = Path(
    "workspace/datasets/dclm-baseline/toktxt/qwen3"
).resolve()
training.save_location = Path(
    "workspace/checkpoints/qwen3-30b-a3b"
).resolve()

training.optimizer = "Adam"
training.scheduler = "CosineAnnealing"
training.max_lr = 3.0e-4
training.min_lr = 1.0e-5
training.warmup_steps = 128
training.max_steps = 10

training.micro_batch_size = 1
training.global_batch_size = 1024
training.sequence_length = 8192

training.moe_load_balance_type = "global-batch"
training.moe_load_balance_coef = 1.0e-3
training.fp8_training = "deep-gemm"

training.seed = 1234
training.save_interval = None

cfg.logging.wandb = LoggingWandbCfg()
cfg.logging.wandb.project = "W2-8-B200-pithtrain-paper-repro"
cfg.logging.wandb.name = "pithtrain-qwen3-30b-a3b-seq8192"

if __name__ == "__main__":
    launch(cfg)
```

The checkpoint was PithTrain's step-0 Torch DCP conversion of
`Qwen/Qwen3-30B-A3B`. We also verified sampled tensors against the original HF
safetensors before training. The data was the Qwen-tokenized
`mlfoundations/dclm-baseline-1.0` corpus prepared through PithTrain, with seed
1234.

## Load-balancing loss

The numbers below are the coefficient-free load-balancing metric reported by
the two trainers. The contribution added to the training objective is the
reported value multiplied by the configured coefficient, `1e-3`.

At the initial Qwen checkpoint:

| Run | First reported load-balancing loss |
|---|---:|
| PithTrain, seq8192, MBS1/GBS1024 | `1.87393872` |
| TorchTitan, seq8192, MBS1/GBS1024 | `1.87173494` |
| TorchTitan fast run, seq8192, MBS8/GBS64 | `1.85406351` |
| TorchTitan fast run, seq4096, MBS14/GBS112 | `1.87440252` |

Thus, at initialization, both implementations report a coefficient-free value
of approximately `1.87`, corresponding to an approximately `0.00187` weighted
contribution.

For completeness, the first five PithTrain values from the seq8192 run were:

```text
1.87393872
1.83284760
1.86968222
1.89346739
1.84645830
```

The five TorchTitan seq8192/MBS8 values were:

```text
1.85406351
1.82937304
1.87255700
1.74992307
1.77304840
```

One clarification regarding the throughput result: our 19.52k TPS/GPU
TorchTitan result used seq8192, BF16, FSDP8, MBS8, GBS64, selective activation
checkpointing, eight loss chunks, compiled model/loss, and a separate
reduce-scatter communicator. It should be treated as a best-recipe comparison,
not a framework-only ablation against PithTrain's PP2/EP4, FP8, MBS1, GBS1024
configuration. At the matched TorchTitan MBS1/GBS1024 geometry, we measured
approximately 12.05k TPS/GPU.

The later load-balancing values from the fast recipes are not expected to
match step-for-step because the GBS, accumulation geometry, precision, and
number of samples consumed per optimizer update differ. 

The public repository will include the matched-data loader, Pith-compatible global-batch auxiliary
loss, exact TorchTitan configurations, per-step logs, asset hashes, and launch
instructions so that these distinctions are reproducible.

Please let me know if there are any additional measurements or configuration
details you would like me to include when I publish the TorchTitan code.

**Comment by DAN-DNA-AND** [comment]
@haok1402 
please find the code here https://github.com/whaletech-ai/Titan-fsdp-Compact-and-Agent-Native-MoE-Training-System.
hope to hear from you soon!

**Comment by haok1402** [comment]
Thanks for the thorough report, @DAN-DNA-AND! I re-ran the sweep on 8×B200, and I think the gap mostly comes down to three things.

First, routing balance. EP is sensitive to load balance in a way FSDP isn't: with experts sharded across ranks, an imbalanced batch leaves some ranks doing more work while the all-to-all waits on the slowest (stragglers), whereas FSDP shards no experts so every rank does equal work regardless of routing. That's why MoE benchmarks force-balance ([NeMo does the same](https://docs.nvidia.com/nemo/megatron-bridge/latest/performance-summary.html#moe-training-note)), and a 1.87 LB loss is fairly skewed (enforced balancing usually sits around 1.0 to 1.2). With `benchmark=True` (force-balanced, dropless), our PP2-DP1-EP4 FP8 run reads 18.20k tok/s/GPU.

Second, precision. Your run used FP8 while TorchTitan used BF16. At the scale of Qwen3-30B-A3B, FP8 isn't our throughput-optimal choice; the quantization overhead can actually make it a bit slower than BF16. Same PP2-DP1-EP4 in BF16 gives 19.74k.

Third, recipe. Your 19.5k is a best-recipe number, which is fair. Our best BF16 layout at matched seq8192/MBS1/GBS1024 is PP1-DP2-EP4 at 20.3k (vs 19.7k for PP2-DP1-EP4 and 19.4k for PP1-DP1-EP8). We can't push DP any further yet since we don't have activation checkpointing (tricky to compose with DualPipeV's manually scheduled overlap), but it's on our todo.

**Comment by haok1402** [comment]
@DAN-DNA-AND Hope this helps clarify where the difference comes from. Please let me know if it does, or if there are any remaining questions or numbers you'd like us to dig into further. Happy to keep the discussion going, and thanks again for the careful comparison.

**Comment by DAN-DNA-AND** [comment]
@haok1402 @MasterJH5574  Thanks for the clarification and for providing the additional recipes.

One clarification on my side: commit `23db182` is the PithTrain revision pinned by the paper, while the FP8 setting comes specifically from the Qwen3-30B-A3B B200 row in Table 4. I reconstructed that row's PP2-DP1-EP4 / seq8192 / FP8 configuration; it is not the default Qwen example shipped at that commit.

<img width="548" height="137" alt="Image" src="https://github.com/user-attachments/assets/fc9e4d83-7b71-4d6f-8e02-2420b8b26f73" />

https://arxiv.org/pdf/2605.31463

The resulting 16–17k tok/s/GPU is consistent with the paper's reported 134.5k aggregate tok/s, or approximately 16.8k tok/s/GPU.

I failed to run  PP2-DP1-EP4 FP8 with bf16 when force-balance is off due to OOM.

With benchmark=True (force-balanced, dropless), it will become very different with the actual training semantics where the paper setup is more reasonable, an actual resume training for a stable qwen ckpt with force balance = off (and it is also the setup for our titan fsdp recipe).

I would be more interested in results from an actual training setup, as megatron can achieve [40k tps/gpu](https://docs.nvidia.com/nemo/megatron-bridge/latest/performance-summary.html) with force balance and mbs=4096 but it cannot be used for actual training at all.

[MoonEP](https://github.com/moonshotAI/moonep) is an interesting EP to solve this imbalance problem with force balance on and it can actually be implemented for real training not just benchmark tps.

I understand PithTrain is a project more for agent efficiency than training performance, but actual training performance is the most appealing factor to most use cases as agent tokens are far cheaper than GPU costs.

Thanks again, and I look forward to activation checkpointing and the other planned features.

**Comment by michaelmanly** [comment]
@DAN-DNA-AND 
The interesting comparison now seems to be the real-training path rather than another force-balanced benchmark number.

Do you have the smallest launch command that reproduces the force-balance-off Qwen3-30B-A3B run on 8×B200?

If you share that plus the pinned config/checkpoint, I can map it into a capped job that runs just enough steps to confirm the setup, record tok/s/GPU + peak memory + load-balance loss, then stop. I can give you the exact command and some credit to try the run.
